app/lib/service/openid/jwt.dart

```diff
@@ -123,7 +114,179 @@ Map<String, dynamic> _decodePart(String part) {
+    } catch (_) {
+      throw FormatException('Unexpected value in JWT payload.');
+    }
+  }
+
+  /// Returns the first key from [keys] which is missing or has `null` value.
+  ///
+  /// Returns `null` if all the [keys] are present with non-null values.
+  String? firstMissingKey(Iterable<String> keys) {
+    for (final key in keys) {
+      final value = this[key];
+      if (value == null) {
+        return key;
+      }
+    }
+    return null;
+  }
+
+  /// Verifies the timestamps with to the provided [now]:
+  /// - [iat] <= [now]
+  /// - [nbf] <= [now]
+  /// - [now] <= [exp]
+  bool verifyTimestamps([DateTime? now]) {
+    now ??= clock.now();
+    if (iat != null && iat!.isAfter(now)) {
+      return false;
+    }
+    if (nbf != null && nbf!.isAfter(now)) {
+      return false;
+    }
+    if (exp != null && exp!.isBefore(now)) {
+      return false;
+    }
+    return true;
+  }
+}
+
+/// Parsed payload with the payload values GitHub sends with the token.
+class GitHubJwtPayload extends JwtPayload {
+  /// timestamp when token was issued
+  @override
+  DateTime get iat => super.iat!;
+
+  /// timestamp before which the token is not valid
+  @override
+  DateTime get nbf => super.nbf!;
+
+  /// timestamp at which the token expires
+  @override
+  DateTime get exp => super.exp!;
+
+  /// user controllable URL identifying the intended audience
+  final String aud;
+
+  /// repository for which the action is running
+  final String repository;
+
+  /// owner of the repository
+  final String repositoryOwner;
+
+  /// name of the event that triggered the workflow
+  final String eventName;
+
+  /// git reference (e.g. branch name or tag name)
+  final String ref;
+
+  /// the kind of git reference given (e.g. "branch")
+  final String refType;
+
+  /// name of the environment used by the job
+  final String environment;
+
+  static const _keys = <String>[
+    'iat',
+    'nbf',
+    'exp',
+    'aud',
+    'repository',
+    'repository_owner',
+    'event_name',
+    'ref',
+    'ref_type',
+    'environment',
+  ];
+
+  GitHubJwtPayload._(Map<String, dynamic> map)
+      : aud = map['aud'] as String,
+        repository = map['repository'] as String,
+        repositoryOwner = map['repository_owner'] as String,
+        eventName = map['event_name'] as String,
+        ref = map['ref'] as String,
+        refType = map['ref_type'] as String,
+        environment = map['environment'] as String,
+        super._(map);
+
+  factory GitHubJwtPayload(JwtPayload payload) {
+    final missing = payload.firstMissingKey(_keys);
+    if (missing != null) {
+      throw FormatException('Missing key: `$missing`.');
+    }
+    return GitHubJwtPayload._(payload);
+  }
+
+  static GitHubJwtPayload? tryParse(JwtPayload payload) {
+    final missing = payload.firstMissingKey(_keys);
+    if (missing != null) {
+      return null;
+    }
+    try {
+      return GitHubJwtPayload(payload);
+    } catch (_) {
+      return null;
+    }
+  }
```

Comment:
Why make it any more complicated than this?
```suggestion
  static GitHubJwtPayload? tryParse(JwtPayload payload) {
    try {
      return GitHubJwtPayload(payload);
    } on FormatException {
      return null;
    }
  }
```

IMO we should only catch `FormatException` because that's the only thing `GitHubJwtPayload()` is allowed to throw (well, if we want to be perfectly right we should document that too).

Reply:
The reason for the catch-all was to catch type-cast exceptions too, which makes the entire initialization easier... but of course we could also verify the type before initializing the field :)